iwyu-check-license-header.py

```diff
@@ -0,0 +1,266 @@
+
+        self.lines = list(content.splitlines())
+        self.hdrindex = find_license_header(self.lines, self.two)
+        self.filename = filename
+        self.errors = []
+
+    @classmethod
+    def parse(_, filename):
+        """ Return an object derived from File to analyze license headers """
+        _, ext = os.path.splitext(filename)
+        if ext == '.py':
+            klass = PythonFile
+        elif ext in ('.h', '.c', '.cc'):
+            klass = CxxFile
+        else:
+            return None
+
+        return klass(filename)
+
+    def has_license_header(self):
+        """ Return True if a license header has been found """
+        return self.hdrindex != -1
+
+    def add_license_header(self):
+        """ Add license header to a file that doesn't have one. """
+        assert not self.has_license_header()
+
+        # Find insertion point
+        for p, line in enumerate(self.lines):
+            # Skip past leading blank lines and hash-bangs.
+            if line and not line.startswith('#!'):
+                break
+
+        # Split the lines around the insertion point
+        if self.lines:
+            before, after = self.lines[:p], self.lines[p:]
+        else:
+            before, after = [], []
+
+        # Rebuild the contents with the license header in the middle
+        lines = before
+        if before and before[-1] != '':
+            lines += ['']
+        lines += make_license_header(self.filename, self.one, self.two)
+        if after and after[0] != '':
+            lines += ['']
+        lines += after
+
+        # Write back out
+        with open(self.filename, 'wb') as fd:
+            fd.write('\n'.join(lines))
+            fd.write('\n')
+
+    def check_license_header(self):
+        """ Check that the header lines follow convention.
+
+        Returns True if everything is OK, otherwise returns False and populates
+        self.errors with all found errors.
+        """
+        if not self.has_license_header():
+            self.file_error('No license header found')
+            return False
+
+        hdrlines = self.lines[self.hdrindex:self.hdrindex+8]
+
+        # First line has the most structure
+        line = hdrlines[0]
+        if len(line) != 80:
+            self.line_error(
+                1, 'Bad header line length (expected: 80, was: %d)' % len(line),
+                " Header line: '%s'" % line)
+
+        m = self.pattern.match(line)
+        if not m:
+            self.line_error(1, 'Bad header line',
+                            " Expected: '%s'" % self.pattern.pattern,
+                            " Actual:   '%s'" % line)
+        else:
+            hfilename = m.group(1)
+            xfilename = os.path.basename(self.filename)
+            if hfilename != xfilename:
+                self.line_error(1, 'Bad header filename',
+                                " Expected:  '%s'" % xfilename,
+                                " Actual:    '%s'" % hfilename)
```

Comment:
This will fail for truncated filenames (longer than 65)

Reply:
Fixed